### PR TITLE
[InstCombine] Fold (X << Y) / (X << Z) -> 1 << Y >> Z

### LINKED ISSUE #68857: [InstCombine] (a+a+a+a)/(a+a) is not optimized
Test case: https://godbolt.org/z/7caGb3avf
```c
int n1;
void func1(int a){

    if(a>0&&a<10){
        n1=(a+a+a+a)/(a+a);
    }
}
```
It appears that `InstCombine` is missing this pattern `(a+a+a+a)/(a+a) => 2`
clang -O3:
```llvm
if.then:
  %a.tr = trunc i32 %a to i8
  %div.lhs.trunc = shl nuw nsw i8 %a.tr, 2
  %div.rhs.trunc = shl nuw nsw i8 %a.tr, 1
  %div12 = udiv i8 %div.lhs.trunc, %div.rhs.trunc
  %div.zext = zext i8 %div12 to i32
  store i32 %div.zext, ptr @n1, align 4
  br label %if.end
```
```asm
func1(int):                              # @func1(int)
        lea     eax, [rdi - 1]
        cmp     eax, 8
        ja      .LBB0_2
        lea     eax, [4*rdi]
        add     dil, dil
        movzx   eax, al
        div     dil
        movzx   eax, al
        mov     dword ptr [rip + n1], eax
.LBB0_2:                                # %if.end
        ret
```

proof: https://alive2.llvm.org/ce/z/5UXwMN

### PR DESCRIPTION
**Pull request by XChy**
Resolve #68857.
Alive2 proofs:
[Proof1](https://alive2.llvm.org/ce/z/_N7cxo)
[Proof2](https://alive2.llvm.org/ce/z/Q58yJM)
Edit:
[Whole proofs](https://alive2.llvm.org/ce/z/A5b85F)

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:999]:
```suggestion
    if (IsSigned && (Shl0->hasNoSignedWrap() && Shl1->hasNoSignedWrap() && Shl1->hasNoUnsignedWrap()))
```
See also: https://github.com/llvm/llvm-project/blob/2ceabf6bdc649f06a316f917a0f508fe10ea7faa/llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp#L976-L980
Alive2: https://alive2.llvm.org/ce/z/sscBF7
Please add a negative test.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:988]: Please add some comments here and add test `sdiv_shl_pair_multi_use`.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

**Comment by XChy** [comment]
Thanks for your review. Here is a more direct fold. [Alive2](https://alive2.llvm.org/ce/z/iDt-wU).

**Comment by dtcxzyw** [comment]
> Thanks for your review. Here is a more direct fold. [Alive2](https://alive2.llvm.org/ce/z/_N7cxo).

Please update the PR title.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:989]:
```suggestion
  // Only when both Op0 and Op1 have uses, this fold produces one extra
  // instruction.
  if (match(Op0, m_Shl(m_Value(X), m_Value(Y))) &&
      match(Op1, m_Shl(m_Specific(X), m_Value(Z))) &&
```
Please also remove duplicate logic that folds `(1 << Y) / (1 << Z) -> 1 << Y >> Z`.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1000]:
```suggestion
          Builder.CreateShl(One, Y, "shl.dividend", /*HasNUW*/ true, /*HasNSW*/ IsSigned);
```

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

**Review by dtcxzyw — APPROVED** [review]
LGTM. Waiting for additional approval from other reviewers.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:997]: These cases also work: https://alive2.llvm.org/ce/z/Q58yJM

So you can do the fold with only two nsw, but can only preserve the nsw flag if either of the shifts is also nuw.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1000]: This name is pointless and will be ignored.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:993]: nit: Move this to inside of the below if statement.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:989]: maybe @nikic cay more, but I'd be okay with dropping the use check to get rid of the `div`. We already increase inst count to drop `div` in other places.

Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:989]: Yeah, I think that would be fine.

**Comment by XChy** [comment]
For udiv, nsw in dividend should be preserved too. [Whole proofs](https://alive2.llvm.org/ce/z/A5b85F)

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:998]:
```suggestion
          /*HasNUW*/ true, /*HasNSW*/ IsSigned ? (Shl0->hasNoUnsignedWrap() || Shl1->hasNoUnsignedWrap()) : Shl0->hasNoSignedWrap());
```
Please add more tests about flag preserving.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/test/Transforms/InstCombine/div-shift.ll
Comment (nikic) [inline llvm/test/Transforms/InstCombine/div-shift.ll:1096]: Isn't this test case the same as the previous one? Did you mean to omit the nuw maybe?

File: llvm/test/Transforms/InstCombine/div-shift.ll
Comment (nikic) [inline llvm/test/Transforms/InstCombine/div-shift.ll:1213]: Can you please also add a negative test where the LHS operands are not the same?

**Comment by nikic** [comment]
This also needs a multi-use test.

File: llvm/test/Transforms/InstCombine/div-shift.ll
Comment (XChy) [inline llvm/test/Transforms/InstCombine/div-shift.ll:1111]: Different LHS test here.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:991]: nit: Would prefer ternary here i.e `isSigned? (Shl0->Nsw && Shl1->Nsw) : (...Nuw)`

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**
